### ISSUE TITLE
Reactive streams

### ISSUE THREAD

**Issue by cbornet** [grpc/grpc#9494]
Are there any plans to support [reactive streams](http://www.reactive-streams.org/) to get non-blocking back-pressure ?

**Comment by ctiller** [comment]
Which language are you interested in?

**Comment by cbornet** [comment]
I'm interested in java and python. Do you mean that it can be supported by languages individually ? I thought it would need protocol adaptations first.

**Comment by ctiller** [comment]
Reactive streams are really just sugar on APIs that we already have in
place. I don't expect any protocol work would be necessary. It's not
possible to write gRPC code without support for backpressure.

**Comment by cbornet** [comment]
But reactive streams expect some cooperation between subscribers and publishers where the subscriber indicates to the publisher how many work unit it has available. So there must be some signaling protocol from the subscriber to the publisher. Maybe it's already possible with gRPC (I don't know the internal very much) but all the stream back pressure implementations I've seen so far are based on blocking a queue when it's full which may cause deadlocks in some situations.

**Comment by ctiller** [comment]
So as far as I know all gRPC implementations have a finite buffer length
for messages on a stream which provides the backpressure you're looking for.

**Comment by cbornet** [comment]
Yes. That's the usual way when doing "push" stream. When the bounded queue if full (when the publisher is faster than the subscriber), you either drop new attempts to write to the queue or block the publisher.
But this has some drawbacks when doing fully non-blocking async architectures (the first being that you block, the second that you may deadlock, ...). That's what reactive streams try to improve by using a "pull" stream pattern.

**Comment by cbornet** [comment]
So it seems HTTP2 flow control on which gRPC is built works that way. Thus you're right, it's only about language implementation.
For Java there seem to be some effort done in that way although it's missing docs and examples.
For python, I don't know if there is something like that. Can you tell ?

**Comment by jroper** [comment]
As the author of the Netty reactive streams implementation (so I know what's involved in implementing reactive streams), I've just taken a cursory glance at the streaming/backpressure APIs that the Java gRPC library offers.  The APIs are not symmetric for sending/receiving, so this will mean two different implementations will need to be created.  Furthermore, the server and client APIs are not shared, though from a streaming/backpressure perspective, they are essentially the same, so each implementation will be duplicated, but that should be straight forward.

For receiving data, it should be very straight forward to wrap the Client/ServerCall in a Publisher, the combination of `start`, `request`, `onMessage`, and `onClose` should map well trivially to the reactive streams `onSubscribe`, `request`, `onNext`, `onComplete` and `onError` methods, so implementing that will only be a handful of lines of essentially delegate methods, with a little bit of logic necessary to decode the `onClose` message into either `onComplete` or `onError`.

For sending data, it's not so trivial. gRPC offers `isReady`, `sendMessage`, and `onReady`, plus completion/error handling. This doesn't map trivially down onto the reactive streams `request` and `onNext` mechanism, so to implement it, some state (outstanding requested messages, with high/low watermarks for triggering more requests) will need to be kept, and concurrency will have to be dealt with, since this state will be modified by both the gRPC and the reactive streams publisher implementations concurrently.  It's not impossible, it's just not trivial.  One thing that would make it a lot easier is if the implementation of `ClientCall` (I haven't looked in depth at `ServerCall`) would expose its underlying executor, this would allow the invocation of callbacks coming from the publisher to the same thread pool with the same serializability guarantees offered by the callbacks executed by gRPC on the listener, and would essentially eliminate the concurrency issues.

**Comment by mushtaq** [comment]
It will be really nice if grpc-java supports Reactive-Streams SPI. Is this being considered for a future release?

**Comment by cbornet** [comment]
See https://github.com/salesforce/grpc-java-contrib/tree/master/rxgrpc for RxJava2 generator plugin with flow control enabled.

**Comment by ncteisen** [comment]
Closing in favor of the RFC

**Comment by cbornet** [comment]
@ncteisen which RFC ?

**Comment by cbornet** [comment]
BTW the reactive bindings are now in https://github.com/salesforce/reactive-grpc and support both Rx-Java 2 and Reactor

**Comment by ncteisen** [comment]
This one: https://github.com/grpc/proposal/pull/33

Does that fulfill this bug? We are doing quick triaging of backlogs today. If this is a separate issue we will leave it open

**Comment by cbornet** [comment]
The RFC you mention is only for C++.
That said it's already totally possible to implement Reactive-Streams over gRPC as is done for Java by reactive-grpc. So I guess this issue can be closed as the rest is language dependant.
